### DISCUSSION TITLE
Variable Mass System Dynamics simulations

### DISCUSSION THREAD

**Discussion by angadhn** [google-deepmind/mujoco#914] [Ideas]
After Friday's workshops (2/3/2023), @yuvaltassa and I had a chat about simulating 6-DoF rocket-type systems in Mujoco, i.e. systems that are characterised by variable mass and variable geometry that then impact the moments of inertia. Yuval also recommended opening a discussion here- after duly going through the tutorials- about the system I wish to simulate that could potentially be added as an example to MuJoCo, which I am happy to do.

# The System
The baseline example I am looking to simulate is described in [this paper](https://link.springer.com/article/10.1007/s11071-018-4485-6), which investigates the attitude dynamics of torque-free variable mass cylinders subjected to four propellant "burns". One of the burns assumes the cylinder's mass is lost uniformly without altering its overall dimensions (i.e., there is only an $\dot{m}$ term which influences the time rates of changes of the moments of inertia); the other three account for geometry changes, as well (e.g., a rocket motor might consume propellant burning from one end of a cylinder to the other so the length of the cylinder reduces along with the mass). The paper assumes there is no gravity ($g=0$) but there are interesting enough behaviours seen across these burn patterns (which we don't need to get into for now). As a result, the governing equations resemble the Euler equations of a single rigid body with some corrective terms for mass variation (Equations 11-13 from said paper are below, for reference).

![image](https://github.com/deepmind/mujoco/assets/1582630/2609eb39-fbbf-413f-adb8-3779b1720649)

Having read through the MuJoCo tutorial, I am circling back to the discussion on Friday and seeking guidance on how one might simulate a variable mass cylinder as described above. I imagine defining the mass, geometry, and (subsequently) the inertias within a simulation is one part of the problem but not sufficient as the governing equations might also need massaging.

However, any advice to get started on this would be much appreciated.

**Comment by saran-t** [comment]
At first glance, I'd say that this can be mostly implemented as a [passive force plugin](https://mujoco.readthedocs.io/en/stable/programming/extension.html) without any need to modify MuJoCo code. However, the variable inertia side of the story might mean that you have to write back into `mjModel` which is supposed to be `const`.

Maybe @yuvaltassa has a better idea, but this might end up being one of those situations where there's no better alternative than just dropping the `const` qualifier.

**Reply by yuvaltassa** [reply]
Yes, I think this would be a case where you should just overwrite the mass and inertia. It should be easy to write a test for this along the lines of: have 2 XMLs, one with full fuel cylinder, one with say half of the fuel is gone, then check that the mass/inertia modification that you're doing to `mjModel` is actually the correct one.

Try this and see if you get stuck. The fields you want to change are `mjModel.{body_mass, body inertia}`, and you probably want the fuel cylinder to be a seperate body, kinematically "welded" (no DoFs) to its parent.
